cdc/cdc/sink/tikv.go

```diff
@@ -209,99 +230,92 @@ type innerBatch struct {
 	OpType model.OpType
 	Keys   [][]byte
 	Values [][]byte
-	TTL    []uint64
-
-	count int
+	TTLs   []uint64
 }
 
 type tikvBatcher struct {
-	Batches        map[model.OpType]*innerBatch
-	BatchesWithTTL map[model.OpType]*innerBatch
-	count          int
-	byteSize       int64
-}
-
-func NewTiKVBatcher() *tikvBatcher {
-	tikvBatcher := &tikvBatcher{
-		Batches:        make(map[model.OpType]*innerBatch),
-		BatchesWithTTL: make(map[model.OpType]*innerBatch),
-	}
-	tikvBatcher.Batches[model.OpTypePut] = &innerBatch{
-		OpType: model.OpTypePut,
-		Keys:   [][]byte{},
-		Values: [][]byte{},
-	}
-	tikvBatcher.Batches[model.OpTypeDelete] = &innerBatch{
-		OpType: model.OpTypeDelete,
-		Keys:   [][]byte{},
-		Values: [][]byte{},
-	}
-	tikvBatcher.BatchesWithTTL[model.OpTypePut] = &innerBatch{
-		OpType: model.OpTypePut,
-		Keys:   [][]byte{},
-		Values: [][]byte{},
-	}
-	return tikvBatcher
+	Batches  []innerBatch
+	count    int
+	byteSize uint64
+	now      uint64
 }
 
 func (b *tikvBatcher) Count() int {
 	return b.count
 }
 
-func (b *tikvBatcher) ByteSize() int64 {
+func (b *tikvBatcher) ByteSize() uint64 {
 	return b.byteSize
 }
 
-func (b *tikvBatcher) Append(entry *model.RawKVEntry) {
-	log.Debug("[TRACE] tikvBatch::Append", zap.Any("event", entry))
+func (b *tikvBatcher) getNow() uint64 {
+	failpoint.Inject("tikvSinkGetNow", func(val failpoint.Value) {
+		now := uint64(val.(int))
+		failpoint.Return(now)
+	})
+	return uint64(time.Now().Unix()) // TODO: use TSO ?
+}
+
+func extractEntry(entry *model.RawKVEntry, now uint64) (opType model.OpType, key []byte, value []byte, ttl uint64) {
+	opType = entry.OpType
+	key = entry.Key
 
-	var batcher *innerBatch
-	var ok bool
-	if entry.ExpiredTs == 0 {
-		batcher, ok = b.Batches[entry.OpType]
-		if !ok {
-			log.Info("Unknow operate type, support put and delete", zap.Int("OpType", int(entry.OpType)))
+	if entry.OpType == model.OpTypePut {
+		// Expired entries have the effect the same as delete, and can not be ignored.
+		// Do the delete.
+		if entry.ExpiredTs > 0 && entry.ExpiredTs <= now {
```

Comment:
If the system time on `tikv-cdc server` has much gap with the `tikv-server`, the data maybe deleted by mistake.  May  pass data to target TiKV cluster, let TiKV to delete the expired data?

Reply:
The raw_put / raw_batch_put of TiKV accept the time left to live other than an absolute expired timestamp. So it's impossible to do so by now.
Unless we add another argument to accept an absolute expired timestamp. Which can be considered to do later.

Reply:
To solve the time gap between `tikv-cdc server` and `tikv-server`, it seems that we need pass `RemainTTL` instead of `ExpireTsUnixSecs`  in changefeed event.

Reply:
Then the `RemainTTL` will depend on the clock of upstream `tikv-server`. The clock issue is still existed.
Actually, the proposal of absolute timestamp in changefeed event is just to avoid clock issues of upstream.